### ISSUE TITLE
False positives for OperatorPrecedence bug pattern

### ISSUE THREAD

**Issue by davido** [google/error-prone#4153]
We are trying to migrate ti the latest rules_java 7.0.6 version in Bazel, and seeing this 4 false positive breakages: [1].

The common pattern seems to be logical condition combined with ternary operator:

```java
public void preferred(String e) {
    this.preferred = e != null && e.equals(email) ? true : null;
  }
```
that error prone suggesting to change like this:

```java
public void preferred(String e) {
    this.preferred = (e != null && e.equals(email)) ? true : null;
  }
```

or:

```java
info.canDelete =
              perm.ref(ref.getName()).testOrFalse(RefPermission.DELETE)
                      && projectState.statePermitsWrite()
                  ? true
                  : null;
```

that error prone suggesting to change like this:

```java
info.canDelete =
              (permissionBackend.currentUser().ref(name).testOrFalse(RefPermission.DELETE)
                      && rsrc.getProjectState().statePermitsWrite())
                  ? true
                  : null;
```

To reproduce, [clone](https://gerrit.googlesource.com/gerrit/) Gerrit Code Review recursively with this CL applied: [2], and run:

```bash
  $> bazel build java/com/google/gerrit/server/...
```

[1] https://gerrit-review.googlesource.com/c/gerrit/+/390116
[2] https://gerrit-review.googlesource.com/c/gerrit/+/387360

**Comment by cushon** [comment]
The change to OperatorPrecedence was https://github.com/google/error-prone/commit/39d288437add5c0b47f78b00d44c158259b86adf

Can you say more about why this is a false positive for you?

The suggest change is behaviour-preserving. The intent is that it makes it easier to read, which is of course subjective, the [Google style guide](https://google.github.io/styleguide/javaguide.html#s4.7-grouping-parentheses) suggest adding parens unless "there is no reasonable chance the code will be misinterpreted without them".

**Comment by davido** [comment]
Thanks for the pointer, makes sense to me. May be worth extending the documentation, after https://github.com/google/error-prone/commit/39d288437add5c0b47f78b00d44c158259b86adf and also mention ternary operator in addition: [1]?

Suggestion:

```
Use grouping parentheses to disambiguate expressions that contain
ternary operator with single || and &&,
both || and &&,
or both shift and arithmetic operators.
```

WDYT?

[1] https://docs.oracle.com/javase/tutorial/java/nutsandbolts/operators.html

**Comment by cushon** [comment]
Updating the docs SG. I wonder about just providing some code examples, instead of trying to use prose to describe the cases it supports (or committing to which exact examples it will suggest parens for).

What do you think about something like

> Use grouping parentheses to disambiguate expressions that could be
misinterpreted.
> 
> For example, consider this:
> 
> ```java
> boolean d = (a && b) || c;",
> boolean e = (a || b) ? c : d;",
> int z = (x + y) << 2;",
> ```
> 
> Instead of this:
> 
> ```java
> boolean r = a && b || c;",
> boolean e = a || b ? c : d;",
> int z = x + y << 2;",
> ```
